articles/102_dependency_groups.md

```diff
@@ -0,0 +1,168 @@
+---
+layout: default
+title: Dependency groups
+permalink: articles/dependency_groups.html
+abstract:
+  This article describes the concept of dependency groups, their use cases, how they can described in the package manifest, and how they can be mapped into binary packages.
+author: '[Dirk Thomas](https://github.com/dirk-thomas)'
+published: true
+categories: Overview
+---
+
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+## Preface
+
+In the ROS ecosystem a ROS package provides information about its dependencies in the manifest.
+The latest format is specified in the [REP 140](http://www.ros.org/reps/rep-0140.html).
+It supports different kinds of dependencies (`build`, `exec`, `test`, `doc`, etc.) but each dependency is mandatory and needs to satisfied on its own.
+
+This article describes the concept of "dependency groups" which could be used for different purposes.
+In terms of terminology a dependency group is a set of dependencies identified by the name of the group.
+For such a group a different semantic can then be applied (instead of just satisfying each dependency on its own).
+
+## Use Cases
+
+The following subsections will describe how dependencies could be used in two different use cases.
+
+### (1) Need at least one "Provider"
+
+A package might want to declare that at least one dependency from a group needs to be present (called `satisfy-at-least-one`).
+Compared to "normal" dependencies it is not required that all dependencies of that group need to be present to satisfy the dependency group.
+
+An example for this is the `rmw_implementation` package.
+Currently it depends on all RMW implementation packages (e.g. `rmw_fastrtps_cpp`, `rmw_connext_cpp`, etc.) in order to ensure that it may only be built after all RMW implementation packages have been built.
+The goal would be that at least one RMW implementation needs to be present in order to build or use the `rmw_implementation` package.
+It would be acceptable if more than one or even all RMW implementations are available but that is not required.
+
+In case multiple package declare to be part of that dependency group the `rmw_implementation` package might want to provide an ordered list of preferred names which should be considered in that order.
+
+### (2) All packages of a specific "type"
+
+Another use case is that a dependency group doesn't need to be explicitly defined by all dependency names (called `satisfy-all`).
+Instead a dependency group is defined by an identifier and the expectation is that all packages which declare that they are being part of that group should be available.
```

Comment:
I am struggling following this. If not by an identifier (and I assume this could be a string here) how do you declare a dependency group in use case 1)? What is the difference between for example a group `rmw_implementations` and a group `message_packages` ?

Reply:
I don't understand your question. Both groups are just defined by the name used in a `group_depend` tag. Any package can then declare to be a member of it. There is no difference between use case 1 and 2 in terms of how the groups are defined - only the criteria how they are satisfied is different.